gdal/frmts/fits/fitsdataset.cpp

```diff
@@ -599,6 +759,780 @@ GDALDataset *FITSDataset::Create( const char* pszFilename,
+
+                    /* Lambert Conformal Conic 1SP / COO projection */
+                    } else if( strcmp(pstr,"COO" ) == 0 ) {
+                        projName.assign("Lambert_Conformal_Conic_1SP_");
+                        oSRS.SetLCC1SP (crval2, crval1, scale, falseEast, falseNorth);
+
+                    /* Orthographic / SIN projection */
+                    } else if( strcmp(pstr,"SIN" ) == 0 ) {
+                        projName.assign("Orthographic_");
+                        oSRS.SetOrthographic(crval2, crval1, falseEast, falseNorth);
+
+                    /* Point Perspective / AZP projection */
+                    } else if( strcmp(pstr,"AZP" ) == 0 ) {
+                        projName.assign("perspective_point_height_");
+                        oSRS.SetProjection(SRS_PP_PERSPECTIVE_POINT_HEIGHT);
+                        /* # appears to need height... maybe center lon/lat */
+
+                    /* Polar Stereographic / STG projection */
+                    } else if( strcmp(pstr,"STG" ) == 0 ) {
+                        projName.assign("Polar_Stereographic_");
+                        oSRS.SetStereographic(crval2, crval1, scale, falseEast, falseNorth);
+                    } else {
+                        CPLError(CE_Failure, CPLE_AppDefined,
+                             "Unknown projection.");
+                    }
+
+                    projName.append(target);
+                    oSRS.SetProjParm(SRS_PP_FALSE_EASTING,0.0);
+                    oSRS.SetProjParm(SRS_PP_FALSE_NORTHING,0.0);
+
+                    oSRS.SetNode("PROJCS",projName.c_str());
+
+                    oSRS.SetGeogCS(GeogName.c_str(), DatumName.c_str(), target, aRadius, invFlattening,
+                        "Reference_Meridian", 0.0, "degree", 0.0174532925199433);
+                }  else {
+                    CPLError(CE_Failure, CPLE_AppDefined,
+                         "Unknown projection.");
+                }
+            }
+        } else {
+            CPLError(CE_Failure, CPLE_AppDefined,
+                 "No CTYPE keywords: no geospatial information available.");
+            status = 0;
+        }
+    }
+
+    CPLAssert(!bMetadataChanged);
+    CPLAssert(!bNoDataChanged);
+
+    bMetadataChanged = false;
+    bNoDataChanged = false;
+
+    bitpix = this->fitsDataType;
+    FITSRasterBand *poBand = cpl::down_cast<FITSRasterBand*>(GetRasterBand(1));
+
+    if (bitpix != TUSHORT && bitpix != TUINT)
+    {
+        fits_read_key(hFITS, TDOUBLE, "BSCALE", &dfScale, nullptr, &status);
+        if( status )
+        {
+            status = 0;
+            dfScale = 1.;
+        }
+        fits_read_key(hFITS, TDOUBLE, "BZERO", &dfOffset, nullptr, &status);
+        if( status )
+        {
+            status = 0;
+            dfOffset = 0.;
+        }
+    } else {
+        poBand->bHaveOffsetScale = true;
+    }
+
+    if( !poBand->bHaveOffsetScale )
+    {
+        poBand->bHaveOffsetScale = true;
```

Comment:
this block should be moved above in the if (bitpix != TUSHORT && bitpix != TUINT) block, if both BSCALE & BZERO have been found